File: src/main/java/org/sopt/user/service/UserService.java
package org.sopt.user.service;

import org.sopt.user.domain.User;
import org.sopt.user.dto.UserCreateRequest;
import org.sopt.user.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class UserService {

	private final UserRepository userRepository;
	private final PasswordEncoder passwordEncoder;

	public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
		this.userRepository = userRepository;
		this.passwordEncoder = passwordEncoder;
	}

	@Transactional
	public Long join(UserCreateRequest request) {
		String encodedPassword = passwordEncoder.encode(request.password);

		User user = new User(request.name, request.email, encodedPassword);

		User savedUser = userRepository.save(user);

		return savedUser.id;
	}
}

File: src/main/kotlin/org/sopt/user/service/UserService.kt
package org.sopt.user.service

import org.sopt.user.domain.User
import org.sopt.user.dto.UserCreateRequest
import org.sopt.user.repository.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class UserService(private val userRepository: UserRepository, private val passwordEncoder: PasswordEncoder) {
    @Transactional
    fun join(request: UserCreateRequest): Long? {
        val encodedPassword = passwordEncoder.encode(request.password)

        val user = User(request.name, request.email, encodedPassword)

        val savedUser = userRepository.save(user)

        return savedUser.id
    }
}
